### PR TITLE
Add Tensor::internal_amp_non_finite_check_and_unscale

### PR DESCRIPTION
**Pull request by danieldk**
This function unscales a tensor and at the same time checks for
infinities.

This is one of the methods needed to implement gradient scaling for
auto mixed precision.

### CODE REVIEW

**Review by LaurentMazare — APPROVED** [review]

**Comment by LaurentMazare** [comment]
Thanks for the PR!

**Comment by danieldk** [comment]
> Thanks for the PR!

Thanks for merging! All the primitives for auto mixed-precision training training should now be in place. I guess it would be nice to have a more user-friendly abstraction for gradient scaling, like PytTorch' `GradScaler`. I something working in one of my projects (still work in progress):

https://github.com/stickeritis/sticker2/blob/mixed-precision/sticker2/src/optimizers/grad_scale.rs

But I don't use the `tch:nn` optimizers, because I needed AdamW with discriminative learning rates/fine tuning and I assume that only one device/GPU is used. But I could look into generalizing it for doing a PR later (unless someone else beats me to it ;)).
